### ISSUE TITLE
`Top` makes recursive type aliases become `Divergent`

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#3699]
### Summary

From what I understand, `Top` on a fully static type should return the type itself, but on recursive type aliases, it instead makes the recursive part become `Divergent`. Not sure if this could show up in actual code.
https://play.ty.dev/5eb8edd3-3e56-4e0e-9533-7110d724b1bd
```py
from typing import reveal_type
from ty_extensions import Top


type A = tuple[A]

def foo(x: A):
    reveal_type(x)  # Revealed type: `tuple[A]`

type B = Top[tuple[B]]

def bar(x: B):
    # Expected type: `tuple[B]`
    reveal_type(x)  # Revealed type: `tuple[Divergent]`
```

### Version

playground 6425b7b13

**Comment by carljm** [comment]
It's a known issue that applying a type mapping to a recursive type can collapse it to `Divergent` at the point of recursion. In this particular case we could probably avoid that.

**Comment by gaborbernat** [comment]
> Not sure if this could show up in actual code.

It does. `isinstance` narrowing against an abstract container synthesizes `Top[...]`, so any recursive alias built on `Sequence`/`Mapping` hits this without anyone writing `Top` by hand.

pypa/build models a `tomllib` value the obvious way and then narrows it (pypa/build#1135):

```python
from __future__ import annotations

from collections.abc import Mapping, Sequence

Covariant = str | int | Sequence['Covariant'] | Mapping[str, 'Covariant']


def narrow(value: Covariant) -> None:
    if isinstance(value, Mapping):
        reveal_type(value)
        reveal_type(value['key'])
```

ty 0.0.59:

```
info[revealed-type]: Revealed type: `(Sequence[Divergent] & Top[Mapping[Unknown, object]]) | (int & Top[Mapping[Unknown, object]]) | Mapping[str, Divergent] | (str & Top[Mapping[Unknown, object]])`
error[invalid-argument-type]: Method `__getitem__` of type `bound method Top[Mapping[Unknown, object]].__getitem__(key: Never, /) -> object` cannot be called with key of type `Literal["key"]` on object of type `Top[Mapping[Unknown, object]]`
```

The `Never` key makes the narrowed mapping unindexable, so reading a key off a narrowed TOML table is an error. pyrefly 1.1.1 reveals `Mapping[str, Covariant]` here and reports nothing.

Declaring the alias with concrete `list`/`dict` avoids the errors, which is what pypa/build did to get green. That trades away real type safety: mypy catches an invariance violation at the assignment site that ty misses, because ty degrades the element type to `Divergent` either way.

**Comment by carljm** [comment]
The diagnostic on `key["value"]` is technically a true positive, and it's not caused by this issue. It is unrelated to the type alias being recursive and unrelated to `Divergent`. It's just another case of #1578. The same issue reproduces with a non-recursive type alias:

```py
from collections.abc import Mapping
from typing import reveal_type

class Area: ...

def narrow(value: Area | Mapping[str, int]) -> None:
    if isinstance(value, Mapping):
        reveal_type(value)
        reveal_type(value["key"])
```

It's a true positive because a subclass of `int` or `str` can also be a `Mapping`, but not necessarily a `Mapping` with `str` keys.

(That said, it is definitely true that type-mappings converting a recursive type alias to `Divergent` (this issue) is something that can and does occur in real code.)
